### PR TITLE
Quick info glyph

### PR DESCRIPTION
**Pull request by vasily-kirichenko**
WIP

image

- [ ] Show proper glyph for each language construct

### CODE REVIEW

**Comment by vasily-kirichenko** [comment]
It's ready

**Comment by forki** [comment]
Amazing. You even fix the details. Hero.

**Comment by KevinRansom** [comment]
@vasily-kirichenko 
The CI sees a Build error:
this should repro it:
build vs
````
  D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\QuickInfoProviderTests.fs(103,52): error FS0001: Type mismatch. Expecting a    'FSharpToolTipText * TextSpan * Glyph'    but given a    'FSharpToolTipText * TextSpan'    The tuples have differing lengths of 3 and 2 [D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\VisualFSharp.Unittests.fsproj]


````

**Comment by vasily-kirichenko** [comment]
@KevinRansom fixed.

**Comment by KevinRansom** [comment]
@vasily-kirichenko 
ty

**Comment by vasily-kirichenko** [comment]
Let me know if I should merge with master (when master is green again).

**Comment by KevinRansom** [comment]
@vasily-kirichenko should these use :

SymbolSearchKind.DoesNotIncludeRightColumn
or
SymbolSearchKind.IncludesRightColumn


https://github.com/Microsoft/visualfsharp/blob/master/vsintegration/src/FSharp.Editor/InlineRename/InlineRenameService.fs#L187

And here also
https://github.com/Microsoft/visualfsharp/blob/master/vsintegration/src/FSharp.Editor/Navigation/FindReferencesService.fs#L70

**Comment by vasily-kirichenko** [comment]
Both should use `IncludesRightColumn`

**Comment by KevinRansom** [comment]
Yay ... I guessed right :-)

**Comment by vasily-kirichenko** [comment]
:) All features use `IncludesRightColumn` except Quick Info, because the tooltip appear too early when you move cursor towards a symbol from right side, like this: `symbol |`, should work like this: `symbol|`

**Comment by vasily-kirichenko** [comment]
I've merged with master.

**Comment by KevinRansom** [comment]
@dotnet-bot test this please

**Comment by cartermp** [comment]
Love it!

Next step is hooking up to a presenter to get type colors in there.  I plan on looking at that over Christmas time.  Well, assuming you don't already have it implemented by then :p

**Comment by Pilchie** [comment]
You don't need a custom presenter - you should look at how Roslyn does it [here](http://source.roslyn.io/Microsoft.CodeAnalysis.EditorFeatures/R/b5e20e00eecee976.html) with `TaggedText`

File: vsintegration/packages.config [hunk 1]
```diff
@@ -18,6 +18,8 @@
   <package id="Microsoft.VisualStudio.Shell.Design" version="14.3.25407" targetFramework="net46" />
   <package id="Microsoft.VisualStudio.Settings.15.0" version="15.0.25824-RC" targetFramework="net46" />
   <package id="Microsoft.VisualStudio.Utilities" version="14.3.25407" targetFramework="net46" />
+  <package id="Microsoft.VisualStudio.Imaging" version="14.3.25407" targetFramework="net46" />
```
Comment (KevinRansom) [inline vsintegration/packages.config]: Should you use:
https://www.nuget.org/packages/Microsoft.VisualStudio.Imaging/15.0.25901-RC
and
https://www.nuget.org/packages/Microsoft.VisualStudio.ImageCatalog/15.0.25824-RC

Comment (KevinRansom) [inline vsintegration/packages.config]: Or perhaps we should get 2.0.0-rc2 versions published

File: vsintegration/src/FSharp.Editor/FSharp.Editor.fsproj [hunk 2]
```diff
@@ -113,6 +117,12 @@
     <Reference Include="Microsoft.VisualStudio.CoreUtility, Version=$(RoslynVSBinariesVersion).0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
       <HintPath>$(FSharpSourcesRoot)\..\packages\Microsoft.VisualStudio.CoreUtility.$(RoslynVSPackagesVersion)\lib\net45\Microsoft.VisualStudio.CoreUtility.dll</HintPath>
     </Reference>
+    <Reference Include="Microsoft.VisualStudio.Imaging, Version=$(RoslynVSBinariesVersion).0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
+      <HintPath>$(FSharpSourcesRoot)\..\packages\Microsoft.VisualStudio.Imaging.$(RoslynVSPackagesVersion)\lib\net45\Microsoft.VisualStudio.Imaging.dll</HintPath>
```
Comment (KevinRansom) [inline vsintegration/src/FSharp.Editor/FSharp.Editor.fsproj]: The package path might not be right.

**Comment by vasily-kirichenko** [comment]
@KevinRansom I'm not sure I understand how VS package versioning set up here. Why these references so complicated? https://github.com/Microsoft/visualfsharp/blob/master/vsintegration/src/FSharp.Editor/FSharp.Editor.fsproj#L92-L133

It looks like all VisualStudio.xxx packages are assumed to have exactly same `RoslynVSBinariesVersion` version:

```xml
 <Reference Include="Microsoft.VisualStudio.Editor, Version=$(RoslynVSBinariesVersion).0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
      <HintPath>$(FSharpSourcesRoot)\..\packages\Microsoft.VisualStudio.Editor.$(RoslynVSPackagesVersion)\lib\net45\Microsoft.VisualStudio.Editor.dll</HintPath>
    </Reference>
```

That's not the case we have currently:

```json
<package id="Microsoft.VisualStudio.LanguageServices" version="2.0.0-rc2" targetFramework="net46" />
  <package id="Microsoft.Composition" version="1.0.27" targetFramework="net46" />
  <package id="Microsoft.VisualStudio.Package.LanguageService.14.0" version="14.3.25407" targetFramework="net46" />
  <package id="Microsoft.VisualStudio.Editor" version="14.3.25407" targetFramework="net46" />
  <package id="Microsoft.VisualStudio.Text.UI" version="14.3.25407" targetFramework="net46" />
  <package id="Microsoft.VisualStudio.Text.UI.Wpf" version="14.3.25407" targetFramework="net46" />
  <package id="Microsoft.VisualStudio.Text.Data" version="14.3.25407" targetFramework="net46" />
  <package id="Microsoft.VisualStudio.Text.Logic" version="14.3.25407" targetFramework="net46" />
  <package id="Microsoft.VisualStudio.Threading" version="14.1.131" targetFramework="net46" />
  <package id="Microsoft.VisualStudio.CoreUtility" version="14.3.25407" targetFramework="net46" />
  <package id="Microsoft.VisualStudio.Shell.Design" version="14.3.25407" targetFramework="net46" />
  <package id="Microsoft.VisualStudio.Settings.15.0" version="15.0.25824-RC" targetFramework="net46" />
  <package id="Microsoft.VisualStudio.Utilities" version="14.3.25407" targetFramework="net46" />
  <package id="Microsoft.VisualStudio.Language.StandardClassification" version="14.3.25407" targetFramework="net46" />
  <package id="Microsoft.VisualStudio.Language.Intellisense" version="14.3.25407" targetFramework="net46" />
  <package id="Microsoft.VisualStudio.Designer.Interfaces" version="1.1.4322" />
```
You suggest to update Imaging to RC, so how it should be done? Some packages are available in version 15.0.25901-RC, some are not.

I found this approach hard to maintain. Why just not use VS UI to manage the packages automagically? 
(However, I've not managed to get any results trying to gather updated packages with it)

**Comment by vladima** [comment]
getting there...
image

**Comment by vasily-kirichenko** [comment]
@vladima Oh, cool!

**Comment by vasily-kirichenko** [comment]
OK, I've ruled out the packages versions, now VisualFSharp.sln builds successfully on clean checkout.

**Comment by dungpa** [comment]
@Pilchie @vladima I tried to colorize signature help by changing https://github.com/Microsoft/visualfsharp/blob/ca2ed9bfbec4710802f3782ce8840ec060618347/vsintegration/src/FSharp.Editor/Completion/SignatureHelp.fs#L182-L206 to use TextTags.Class, TextTags.Keyword, etc. However, F# signature helps still have no color. Any hint on creating a signature help presenter for F#?

**Comment by vasily-kirichenko** [comment]
@dungpa this looks relevant http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices.Next/FindReferences/StreamingFindUsagesPresenter.DocumentLocationEntry.cs,105

**Comment by dungpa** [comment]
I think it's relevant but should not be necessary. The signature help is colorized at http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/Intellisense/SignatureHelp/Presentation/SignatureHelpClassifier.cs,50

The signature help classifier is restricted to C#/VB at http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/Intellisense/SignatureHelp/Presentation/SignatureHelpClassifierProvider.cs,13. What is the easiest way to enable this for F# signature help? /cc @Pilchie @vladima

**Comment by dungpa** [comment]
Ok, I found a way in https://github.com/Microsoft/visualfsharp/pull/2034

**Comment by vasily-kirichenko** [comment]
Tests fail because VisualFSharp.unittests.dll cannot be loaded:

image

Any ideas how to spot the wrong/missing assembly?

I think that the way packages are managed in this repository is unhealthy. What about switching to Paket?

/cc @forki

**Comment by vasily-kirichenko** [comment]
OK, I reverted the big changes (VS packages ver. 14 -> 15) and it's now simply adds a single `Microsoft.VisualStudio.Imaging` ver. 14. Same error on loading the test dll :(

**Comment by vasily-kirichenko** [comment]
I'm stuck with this :(

**Comment by KevinRansom** [comment]
@vasily-kirichenko 
I can take a look mate.

**Comment by vasily-kirichenko** [comment]
@KevinRansom thanks!

**Comment by vasily-kirichenko** [comment]
@KevinRansom `Microsoft.VisualStudio.Imaging` is here for a single reason - we use `CryspImage` type from it. I've ported this code http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/Intellisense/QuickInfo/DeferredContent/SymbolGlyphDeferredContent.cs,26 as is, if it helps.

BTW, I had to write this unsafe casting:
```fsharp
let moniker = GlyphExtensions.GetImageMoniker(glyph)
let image = new CrispImage()
image.Moniker <- (box moniker) :?> _
```
because `ImageMoniker` returned by the extension method is different from `CryspImage.Moniker` expects (both are identical simple structures, so it works). Maybe it's a hint of why this type loading error occurs.

**Comment by KevinRansom** [comment]
@vasily-kirichenko @Pilchie @brettfo 

This happens because the Roslyn dll's do not have InternalsVisible enabled for VisualFSharpTests.dll  

https://github.com/Microsoft/visualfsharp/blob/master/vsintegration/src/FSharp.Editor/QuickInfo/QuickInfoProvider.fs#L73

This guy's async returns a value:
'Microsoft.FSharp.Core.FSharpFunc`2[System.Tuple`3[
    Microsoft.FSharp.Compiler.SourceCodeServices.FSharpToolTipText,
    Microsoft.CodeAnalysis.Text.TextSpan,Microsoft.CodeAnalysis.Glyph],System.String]'. 

Microsoft.CodeAnalysis.Glyph is internal and thus can not be seen from the test dll.

Which means it can't be used in test code unless the test code has Internals Visible enabled for it:

Either:
1.Rewrite to completely isolate Microsoft.CodeAnalysis.Glyph  we actually are mostly there already 
or ...
2. Modify Roslyn to Make Glyph public 
or ...
3.  set internals visible to our test assembly.

**Comment by vasily-kirichenko** [comment]
Thanks! Will try #1.

**Comment by vasily-kirichenko** [comment]
@KevinRansom fixed. I can load and run the tests locally. Phew, that was a tricky nasty stuff. Thanks!

**Comment by vasily-kirichenko** [comment]
@dotnet-bot test this please

**Comment by KevinRansom** [comment]
Hmmm that failure is absolutely not expected.
@dotnet-bot test this please

**Comment by vasily-kirichenko** [comment]
@vladima has almost finished his work on colorful quick info tooltips here https://github.com/Microsoft/visualfsharp/compare/master...vladima:vladima/classification which operate on a slightly higher level of abstraction than this PR and it's incompatible with it anyway. So I'm closing this in favor of https://github.com/vladima/visualfsharp/pull/1 (or whatever approach @vladima will choose for showing the glyph). Currently it looks like this:

image

**Comment by KevinRansom** [comment]
Okay ... thanks.

**Comment by vasily-kirichenko** [comment]
@KevinRansom sorry for wasting your time :(

**Comment by KevinRansom** [comment]
@vasily-kirichenko 
It's never a waste of my time mate ... and besides, even if it were (which it wasn't) .... you paid me back many thousands of times with all of your awesome IDE PR's.  I reckon I owe you tons.

Kevin

**Pull request closed without merging**